File: merge_sort.py
def merge_sort(a):
    if len(a) == 1:
        return a

    sorted_l = merge_sort(a[:len(a)/2])
    sorted_r = merge_sort(a[len(a)/2:])
    merged = merge(sorted_l,sorted_r)
    return merged

def merge(l, r):
    m = []
    l_i = 0
    r_i = 0

    for i in range(len(l) + len(r)):
        if l_i < len(l) and r_i < len(r):
            if l[l_i] <= r[r_i]:
                m.append(l[l_i])
                l_i += 1
            else:
                m.append(r[r_i])
                r_i += 1
        elif l_i < len(l):
            m.append(l[l_i])
            l_i += 1
        else:
            m.append(r[r_i])
            r_i += 1
    return m

print merge_sort([5,4,3,2,1,0])            

File: merge_sort.rb
require 'pp'

def merge_sort(a)
  len = a.length
  if len == 1
    return a
  end

  sorted_l = merge_sort(a[0...(len / 2)])
  sorted_r = merge_sort(a[(len / 2)...len])
  merged = merge(sorted_l, sorted_r)
  return merged
end

def merge(l, r)
  m = Array.new
  l_i = 0
  r_i = 0
  0.upto(l.length + r.length - 1) do |i|
    if l_i < l.length && r_i < r.length
      if l[l_i] <= r[r_i]
        m[i] = l[l_i]
        l_i += 1
      else
        m[i] = r[r_i]
        r_i += 1
      end
    elsif l_i < l.length
        m[i] = l[l_i]
        l_i += 1
    else
        m[i] = r[r_i]
        r_i += 1
    end
  end
  return m
end

pp merge_sort([5,4,3,2,1,0])
